### PR TITLE
Use proto rules API instead of struct; Moved as much as possible to promclient; Added rulesAPI RPC to sidecar.

### PR DESCRIPTION
**Pull request by bwplotka**
Signed-off-by: Bartlomiej Plotka <bwplotka@gmail.com>

cc @s-urbaniak

### CODE REVIEW

**Comment by bwplotka** [comment]
This should be ready to go @s-urbaniak!

**Comment by bwplotka** [comment]
PTAL

File: pkg/promclient/promclient.go [hunk 1]
```diff
@@ -466,77 +486,142 @@ func convertScalarJSONToVector(scalarJSONResult json.RawMessage) (model.Vector,
...
+	u.Path = path.Join(u.Path, "/api/v1/series")
+	q := u.Query()
+
+	matcher, err := matchersToString(matchers)
 	if err != nil {
-		return errors.Wrapf(err, "perform GET request against %s", u.String())
+		return nil, errors.Wrap(err, "invalid matchers")
 	}
-	defer runutil.ExhaustCloseWithLogOnErr(logger, resp.Body, "metrics body")
+	fmt.Println(matcher)
```
Comment (yeya24) [inline pkg/promclient/promclient.go]: Remove this?

Comment (bwplotka) [reply pkg/promclient/promclient.go]: done

File: pkg/promclient/promclient.go [hunk 2]
```diff
@@ -336,18 +376,13 @@ func (c *Client) QueryInstant(ctx context.Context, base *url.URL, query string,
...
+	span, ctx := tracing.StartSpan(ctx, "/prom_query_instant HTTP[client]")
+	defer span.Finish()
 
-	req = req.WithContext(ctx)
-
-	resp, err := c.httpClient.Do(req)
+	body, _, err := c.get2xx(ctx, &u)
 	if err != nil {
-		return nil, nil, errors.Wrapf(err, "perform GET request against %s", u.String())
+		return nil, nil, err
```
Comment (s-urbaniak) [inline pkg/promclient/promclient.go]: no wrapping is necessary any more?

Comment (bwplotka) [reply pkg/promclient/promclient.go]: correct

**Comment by s-urbaniak** [comment]
wow, this is huge :astonished:
i am a bit overwhelmed, but i see `RulesServer` now being registered, so i am super happy :tada: :blush:

**Comment by bwplotka** [comment]
Happy to have a quick pair-review @s-urbaniak if needed

File: pkg/testutil/testpromcompatibility/api_compatibility.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+	Labels      labels.Labels `json:"labels"`
+	Annotations labels.Labels `json:"annotations"`
+	State       string        `json:"state"`
+	ActiveAt    *time.Time    `json:"activeAt,omitempty"`
+	Value       string        `json:"value"`
+
+	PartialResponseStrategy string `json:"partialResponseStrategy"`
+}
+
+type AlertingRule struct {
```
Comment (bwplotka) [inline pkg/testutil/testpromcompatibility/api_compatibility.go]: Upstream make it public?

**Comment by s-urbaniak** [comment]
lgtm :+1:

**Comment by s-urbaniak** [comment]
looking at the CI failure :eyes:

**Comment by s-urbaniak** [comment]
yay, I can repro locally, let me try to fix this so I'll learn :construction_worker_man:

File: pkg/promclient/promclient.go (hunk 2 above)
Comment (bwplotka) [reply pkg/promclient/promclient.go]: https://github.com/thanos-io/thanos/pull/2277

**Comment by bwplotka** [comment]
Are you on it @s-urbaniak ? If not, I can fix

**Comment by bwplotka** [comment]
Fixed hopefully.

**Comment by s-urbaniak** [comment]
@bwplotka i was on it, but my fix was related to the test itself, i will continue from here for the remaining failures.

**Comment by bwplotka** [comment]
Thanks @s-urbaniak for help! 

Let's move it forward. 

We still need:
* e2e tests
* proxy implementation
* ruler implementation

**Comment by s-urbaniak** [comment]
@bwplotka on it :+1:

**Pull request merged**